blaze-client/src/test/scala/org/http4s/client/blaze/ClientTimeoutSpec.scala

```diff
@@ -156,5 +157,29 @@ class ClientTimeoutSpec extends Http4sSpec {
 
       c.fetchAs[String](FooRequest).unsafeRunSync must_== "done"
     }
+
+    // Regression test for: https://github.com/http4s/http4s/issues/2386
+    // and https://github.com/http4s/http4s/issues/2338
+    "Eventually timeout on connect timeout" in {
+      val manager = ConnectionManager.basic[IO, BlazeConnection[IO]]({ _ =>
+        // In a real use case this timeout is under OS's control (AsynchronousSocketChannel.connect)
+        IO.sleep(2.seconds) *> IO.raiseError[BlazeConnection[IO]](new IOException())
+      })
+      val c = BlazeClient.makeClient(
+        manager = manager,
+        responseHeaderTimeout = Duration.Inf,
+        idleTimeout = Duration.Inf,
+        requestTimeout = 1.second,
+        scheduler = tickWheel,
+        ec = testExecutionContext
+      )
+
+      // if the 5.seconds timeout is hit, it's a NoSuchElementException,
+      // if the requestTimeout = 1.second is hit then it's a TimeoutException
+      // if establishing connection fails first then it's an IOException
+
+      // The expected behaviour is that the requestTimeout will happen first, but fetchAs will additionally wait for the IO.sleep(2.seconds) to complete.
+      c.fetchAs[String](FooRequest).unsafeRunTimed(5.seconds).get must throwA[TimeoutException]
```

Comment:
Per your PR's description:

>  The requestTimeout still isn't enforced, instead we wait for an OS level timeout of an attempt to establish TCP connection, but at least the request fails eventually, instead of waiting indefinitely.

Why wouldn't an `IOException` be thrown since "the requestTimeout still isn't enforced?"

Reply:
In short: The `requestTimeout` is **why** the request fails, so that's what is returned., but the value of the `requestTimeout` is not **when** the request completes.

In detail - here's what happens:
- sending a request (including establishing connection), and the requestTimeout are started in parallel racing with each other
- after 1 second the requestTimeout is hit, it wins the race. The `racePair` completes with TimeoutException (https://github.com/http4s/http4s/blob/3bed1e6088395cbd8b291f99172d68433b9b8c31/blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala#L147) , 
- fiber.cancel is called (https://github.com/http4s/http4s/blob/3bed1e6088395cbd8b291f99172d68433b9b8c31/blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala#L159) attempting to cancel the proces of sending the request. At that time that process is still during the `acuire` part of a `Bracket` which is uninterruptable. So the `fiber.cancel` waits until the aquire completes and then interrupts the `Bracket`.
- after 2 seconds from the start of the test, the aquire completes with  an IOException. The `fiber.cancel`completes and gets flatMapped with `F.raiseError(t)`, `t` being the original TimeoutException. That becomes the final result of the request.

Reply:
Thanks for that detailed explanation, Rafał!